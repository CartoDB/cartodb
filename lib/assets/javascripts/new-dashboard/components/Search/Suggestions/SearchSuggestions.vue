<template>
  <section class="suggestions" :class="{ 'suggestions--open': isOpen }">
    <router-link
      :to="{ name: searchRoute, params: searchRouteParameters }"
      class="suggestions__header is-caption text"
      :class="{ 'suggestions__header--loading': isFetching }"
      v-if="query"
      @click.native="onPageChange">
      {{ query }} <span v-if="!isFetching">- {{ searchResults.total_entries }} results</span>
    </router-link>

    <ul v-if="searchResults" class="suggestions__content">
      <li v-for="visualization in searchResults.visualizations" :key="visualization.id">
        <SearchSuggestionsItem :item="visualization" @itemClick="onPageChange" />
      </li>
    </ul>
  </section>
</template>

<script>
import _ from 'underscore';
import SearchSuggestionsItem from './SearchSuggestionsItem';

export default {
  name: 'SearchSuggestions',
  components: {
    SearchSuggestionsItem
  },
  props: {
    query: String,
    isOpen: {
      type: Boolean,
      default: false
    }
  },
  created () {
    this.fetchSuggestionsDebounced = _.debounce(this.fetchSuggestions.bind(this), 500);
  },
  data () {
    return {
      isFetching: true,
      searchResults: []
    };
  },
  watch: {
    query (newQuery) {
      if (newQuery === '') {
        this.isFetching = false;
        this.searchResults = [];
        return;
      }

      this.isFetching = true;
      this.fetchSuggestionsDebounced();
    }
  },
  computed: {
    isSearchingTags () {
      return this.query.includes(':');
    },
    searchRoute () {
      if (this.isSearchingTags) {
        return 'tagSearch';
      }

      return 'search';
    },
    searchRouteParameters () {
      if (this.isSearchingTags) {
        return { tag: this.query.substring(1) };
      }

      return { query: this.query };
    },
    queryParameters () {
      const queryParameters = {
        types: 'derived,table',
        per_page: 4
      };

      if (this.isSearchingTags) {
        queryParameters.tags = this.query.substring(1);
      }

      if (!this.isSearchingTags) {
        queryParameters.q = this.query;
      }

      return queryParameters;
    }
  },
  methods: {
    fetchSuggestions () {
      this.$store.state.client.getVisualization('',
        this.queryParameters,

        (err, _, data) => {
          this.isFetching = false;

          if (err) {
            return;
          }

          this.searchResults = data;
        }
      );
    },
    onPageChange () {
      this.$emit('pageChange');
    }
  }
};
</script>

<style lang="scss" scoped>
@import 'stylesheets/new-dashboard/variables';

.suggestions {
  visibility: hidden;
  position: absolute;
  z-index: 2;
  top: 100%;
  left: 0;
  width: calc(100% - 16px);
  margin-top: 8px;
  border: 1px solid $light-grey;
  border-radius: 2px;
  opacity: 0;
  background-color: #FFF;
  pointer-events: none;

  &.suggestions--open,
  &:hover {
    visibility: visible;
    opacity: 1;
    pointer-events: auto;
  }
}

.suggestions__header {
  display: block;
  position: relative;
  width: 100%;
  padding: 16px 16px 16px 38px;
  overflow: hidden;
  border-bottom: 1px solid $grey;
  color: $text-color;
  text-overflow: ellipsis;
  white-space: nowrap;

  &:hover {
    background-color: rgba($primary-color, 0.05);
    color: #1785FB;
    text-decoration: none;
  }

  &::before {
    content: '';
    position: absolute;
    top: 50%;
    left: 12px;
    width: 18px;
    height: 18px;
    transform: translate3d(0, -50%, 0);
    background-image: url("../../../assets/icons/navbar/dropdown/loupe-dropdown.svg");
    background-repeat: no-repeat;
    background-position: center;
  }

  &.suggestions__header--loading {
    padding-right: 38px;

    &::after {
      content: '';
      position: absolute;
      top: 50%;
      right: 12px;
      width: 18px;
      height: 18px;
      transform: translate3d(0, -50%, 0);
      background-image: url("../../../assets/icons/common/loading.svg");
      background-repeat: no-repeat;
      background-position: center;
    }
  }
}
</style>
