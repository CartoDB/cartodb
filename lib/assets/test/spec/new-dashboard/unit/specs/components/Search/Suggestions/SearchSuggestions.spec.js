import { shallowMount } from '@vue/test-utils';
import SearchSuggestions from 'new-dashboard/components/Search/Suggestions/SearchSuggestions';
import visualizationData from '../../../fixtures/visualizations';

describe('SearchSuggestions.vue', () => {
  describe('Rendering', () => {
    it('should render correctly', () => {
      const searchSuggestions = createSearchSuggestions();
      expect(searchSuggestions).toMatchSnapshot();
    });

    it('should render isFetching state', () => {
      const searchSuggestions = createSearchSuggestions({
        data () {
          return { isFetching: true };
        }
      });

      expect(searchSuggestions).toMatchSnapshot();
    });
  });

  it('should define fetchSuggestionsDebounced', () => {
    const searchSuggestions = createSearchSuggestions();
    expect(searchSuggestions.vm.fetchSuggestionsDebounced).toBeDefined();
  });

  describe('Computed Properties', () => {
    describe('isSearchingTags', () => {
      it('should return true if query includes :', () => {
        const searchSuggestions = createSearchSuggestions({
          propsData: { query: ':tagname' }
        });
        expect(searchSuggestions.vm.isSearchingTags).toBe(true);
      });

      it("should return false if query doesn't include :", () => {
        const searchSuggestions = createSearchSuggestions({
          propsData: { query: 'query text' }
        });
        expect(searchSuggestions.vm.isSearchingTags).toBe(false);
      });
    });

    describe('searchRoute', () => {
      it('should return tagSearch if isSearchingTags is true', () => {
        const searchSuggestions = createSearchSuggestions({
          computed: {
            isSearchingTags () { return true; }
          }
        });
        expect(searchSuggestions.vm.searchRoute).toBe('tagSearch');
      });

      it('should return search if isSearchingTags is false', () => {
        const searchSuggestions = createSearchSuggestions();
        expect(searchSuggestions.vm.searchRoute).toBe('search');
      });
    });

    describe('searchRouteParameters', () => {
      it('should return object with tag property if isSearchingTags is true', () => {
        const searchSuggestions = createSearchSuggestions({
          propsData: { query: ':tagname' },
          computed: {
            isSearchingTags () { return true; }
          }
        });
        expect(searchSuggestions.vm.searchRouteParameters).toEqual({
          tag: 'tagname'
        });
      });

      it('should return object with query property if isSearchingTags is false', () => {
        const searchSuggestions = createSearchSuggestions({
          computed: {
            isSearchingTags () { return false; }
          }
        });
        expect(searchSuggestions.vm.searchRouteParameters).toEqual({
          query: 'Query text'
        });
      });
    });

    describe('queryParameters', () => {
      it('should return proper query parameters if isSearchingTags is true', () => {
        const searchSuggestions = createSearchSuggestions({
          propsData: { query: ':tagname' },
          computed: {
            isSearchingTags () { return true; }
          }
        });
        expect(searchSuggestions.vm.queryParameters).toEqual({
          types: 'derived,table',
          per_page: 4,
          tags: 'tagname'
        });
      });

      it('should return proper query parameters if isSearchingTags is false', () => {
        const searchSuggestions = createSearchSuggestions({
          computed: {
            isSearchingTags () { return false; }
          }
        });
        expect(searchSuggestions.vm.queryParameters).toEqual({
          types: 'derived,table',
          per_page: 4,
          q: 'Query text'
        });
      });
    });
  });

  describe('Methods', () => {
    describe('fetchSuggestions', () => {
      it('should set search results via API', () => {
        const getVisualizationSpy = jest.fn((vizUrl, params, callback) => {
          callback(null, null, visualizationData);
        });

        const state = {
          client: { getVisualization: getVisualizationSpy }
        };

        const queryParameters = {
          per_page: 4,
          q: 'Query text',
          types: 'derived,table'
        };

        const searchSuggestions = createSearchSuggestions({
          mocks: { $store: { state } },
          data () {
            return {
              isFetching: true,
              searchResults: []
            };
          }
        });

        searchSuggestions.vm.fetchSuggestions();

        expect(searchSuggestions.vm.searchResults).toEqual(visualizationData);
        expect(getVisualizationSpy).toHaveBeenCalledWith('', queryParameters, expect.any(Function));
      });

      it('should not set search results via API if request is errored', () => {
        const error = { error: 'API Error' };
        const getVisualizationSpy = jest.fn((vizUrl, params, callback) => {
          callback(error, null, null);
        });

        const state = {
          client: { getVisualization: getVisualizationSpy }
        };

        const searchSuggestions = createSearchSuggestions({
          mocks: { $store: { state } },
          data () {
            return {
              isFetching: true,
              searchResults: []
            };
          }
        });

        searchSuggestions.vm.fetchSuggestions();

        expect(searchSuggestions.vm.searchResults).toEqual([]);
      });
    });

    describe('onPageChange', () => {
      const searchSuggestions = createSearchSuggestions();
      searchSuggestions.vm.onPageChange();
      expect(searchSuggestions.emitted('pageChange')).toBeTruthy();
    });
  });

  describe('Watch', () => {
    describe('query', () => {
      it('should fetch suggestions when query changes', done => {
        const searchSuggestions = createSearchSuggestions();
        spyOn(searchSuggestions.vm, 'fetchSuggestionsDebounced');

        searchSuggestions.vm.query = 'New query';

        searchSuggestions.vm.$nextTick(() => {
          expect(searchSuggestions.vm.isFetching).toBe(true);
          expect(searchSuggestions.vm.fetchSuggestionsDebounced).toHaveBeenCalled();
          done();
        });
      });

      it('should reset fetching state and results when query changes and is empty', done => {
        const searchSuggestions = createSearchSuggestions();
        spyOn(searchSuggestions.vm, 'fetchSuggestionsDebounced');

        searchSuggestions.vm.query = '';

        searchSuggestions.vm.$nextTick(() => {
          expect(searchSuggestions.vm.isFetching).toBe(false);
          expect(searchSuggestions.vm.searchResults).toEqual([]);
          expect(searchSuggestions.vm.fetchSuggestionsDebounced).not.toHaveBeenCalled();
          done();
        });
      });
    });
  });
});

function createSearchSuggestions (options) {
  return shallowMount(SearchSuggestions, {
    propsData: {
      query: 'Query text',
      isOpen: true
    },
    data () {
      return {
        isFetching: false,
        searchResults: visualizationData
      };
    },
    ...options
  });
}
