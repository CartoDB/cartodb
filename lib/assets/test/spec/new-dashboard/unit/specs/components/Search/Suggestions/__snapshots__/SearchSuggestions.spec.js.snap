// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSuggestions.vue Rendering should render correctly 1`] = `
<section class="suggestions suggestions--open">
  <router-link to="[object Object]" class="suggestions__header is-caption text">
    Query text <span>- 1 results</span></router-link>
  <ul class="suggestions__content">
    <li>
      <searchsuggestionsitem-stub item="[object Object]"></searchsuggestionsitem-stub>
    </li>
    <li>
      <searchsuggestionsitem-stub item="[object Object]"></searchsuggestionsitem-stub>
    </li>
  </ul>
</section>
`;

exports[`SearchSuggestions.vue Rendering should render isFetching state 1`] = `
<section class="suggestions suggestions--open">
  <router-link to="[object Object]" class="suggestions__header is-caption text suggestions__header--loading">
    Query text
    <!---->
  </router-link>
  <ul class="suggestions__content"></ul>
</section>
`;
